### PR TITLE
[BugFix]handle nullable for array_predicate

### PR DESCRIPTION
**Pull request by before-Sunrise**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Problem Summary(Required) ：

array_predicate need handle situation like nullable_col(array_col) == nullable(array_col
## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
- [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Review by murphyatwork — DISMISSED** [review]

**Review by imay — COMMENTED** [review]
You can write some sql test which is very convenient

**Review by kangkaisen — DISMISSED** [review]

File: be/src/exprs/binary_predicate.cpp [hunk 1]
```diff
@@ -117,8 +117,18 @@ class ArrayPredicate final : public Predicate {
     StatusOr<ColumnPtr> evaluate_checked(ExprContext* context, Chunk* ptr) override {
         ASSIGN_OR_RETURN(auto l, _children[0]->evaluate_checked(context, ptr));
         ASSIGN_OR_RETURN(auto r, _children[1]->evaluate_checked(context, ptr));
-        auto lhs_arr = std::static_pointer_cast<ArrayColumn>(l);
-        auto rhs_arr = std::static_pointer_cast<ArrayColumn>(r);
+
+        if (l->only_null() || r->only_null()) {
```
Comment (satanson) [inline be/src/exprs/binary_predicate.cpp:121]: for null-safe eq, here must return all trues.

File: test/sql/test_array/T/test_array [hunk 2]
```diff
@@ -1,2 +1,26 @@
...
+    (1,[22, 11, 33],[22, 11, 33],[22, 11, 33],[22, 11, 33]),
+    (2,[22, 11, 33],[22, 11, 33],[22, 11, 33],[22, 11, 33]),
+    (3,[22, 11, 33],[22, 11, 33],[22, 11, 33],[22, 11, 33]);
+
+insert into array_data_type(c1,c3,c5) values (4,[22, 11, 33],[22, 11, 33]);
+
+select c2 = c3 from array_data_type;
+select c3 = c2 from array_data_type;
+select c2 = c4 from array_data_type;
+select c3 = c5 from array_data_type;
```
Comment (satanson) [inline test/sql/test_array/T/test_array:26]: add test cases like this:
```
insert into array_data_type (c1, c2, c3, c4,c5) values 
    (1,NULL,NULL,[22, 11, 33],[22, 11, 33]),
    (2,[22, 11, 33],[22, 11, 33],[22, 11, 33],[22, 11, 33]),
    (3,NULL,[22, 11, 33],[22, 11, 33],[22, 11, 33]);
```

select c2 <=> c3 from array_data_type;

File: be/src/exprs/binary_predicate.cpp [hunk 3]
```diff
@@ -117,8 +117,18 @@ class ArrayPredicate final : public Predicate {
...
+            return ColumnHelper::create_const_null_column(l->size());
+        }
+
+        const ColumnPtr& data1 = FunctionHelper::get_data_column_of_nullable(l);
+        const ColumnPtr& data2 = FunctionHelper::get_data_column_of_nullable(r);
+
+        DCHECK(data1->is_array());
+        DCHECK(data2->is_array());
+        auto lhs_arr = std::static_pointer_cast<ArrayColumn>(data1);
+        auto rhs_arr = std::static_pointer_cast<ArrayColumn>(data2);
```
Comment (fzhedu) [inline be/src/exprs/binary_predicate.cpp:131]: use down_case<>()

File: be/src/exprs/binary_predicate.cpp [hunk 4]
```diff
@@ -117,8 +117,18 @@ class ArrayPredicate final : public Predicate {
...
         ASSIGN_OR_RETURN(auto r, _children[1]->evaluate_checked(context, ptr));
-        auto lhs_arr = std::static_pointer_cast<ArrayColumn>(l);
-        auto rhs_arr = std::static_pointer_cast<ArrayColumn>(r);
+
+        if (l->only_null() || r->only_null()) {
+            return ColumnHelper::create_const_null_column(l->size());
+        }
+
+        const ColumnPtr& data1 = FunctionHelper::get_data_column_of_nullable(l);
+        const ColumnPtr& data2 = FunctionHelper::get_data_column_of_nullable(r);
```
Comment (fzhedu) [inline be/src/exprs/binary_predicate.cpp:126]: need `unpack_and_duplicate_const_column`?

File: test/sql/test_array/T/test_array [hunk 5]
```diff
@@ -1,2 +1,26 @@
...
+    PRIMARY KEY(c1) 
+    DISTRIBUTED BY HASH(c1) 
+    buckets 1 
+    PROPERTIES ("replication_num" = "1");
+
+insert into array_data_type (c1, c2, c3, c4,c5) values 
+    (1,[22, 11, 33],[22, 11, 33],[22, 11, 33],[22, 11, 33]),
+    (2,[22, 11, 33],[22, 11, 33],[22, 11, 33],[22, 11, 33]),
+    (3,[22, 11, 33],[22, 11, 33],[22, 11, 33],[22, 11, 33]);
+
```
Comment (LiShuMing) [inline test/sql/test_array/T/test_array:20]: Add some null values to cover your codes?
```
if (l->only_null() || r->only_null()) {
```

File: be/src/exprs/binary_predicate.cpp (hunk 1 above)
Comment (before-Sunrise) [reply be/src/exprs/binary_predicate.cpp:121]: added another predicate to deal with null safe eq

**Review by kangkaisen — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**